### DISCUSSION TITLE
Cast to double

### DISCUSSION THREAD

**Discussion by condor-cp** [coin-or/CppAD#107] [Q&A]
Hello,
I'm using CppAD and I wonder if there is a specific reason why CppAD::AD doesn't implement the "cast to double" function.

A guess of what it could look like : 
    operator double() const { return CppAD::Value(CppAD::Var2Par(this)); }
    
It would simplify my developpement to have such a function.

Thanks a lot,
Cédric

**Comment by bradbell** [comment]
If you look at the documentation for the Value function
https://coin-or.github.io/CppAD/doc/value.htm
you will see  that you can convert and AD<double> to a double so long as the AD<double> does not depend on the independent variables.

If you want to convert variable, or dynamic parameter to double you must first convert it to a parameter using
https://coin-or.github.io/CppAD/doc/var2par.htm
This way you are explicitly showing that the dependence on the independent variables and dynamic parmaerts is not included.

Perhaps I should add more to the Var2Par documentation stating that dynamic parameters can also be converted.

**Reply by condor-cp** [reply]
Ok my question lacked precision.
I understand that I always can explicitly with the Var2Par + Value functions.
I would need to **implictly** convert an AD to double like so :
```
AD foo = 1.0;
double bar = foo; 
```

I need this because I have a templated function such as the example function under, and I don't want to add explicit conversion in the function in the case the SCALAR is AD.
```
template<typename SCALAR>
double foo(SCALAR bar)
{
    return double(bar);
}
```

This is why I suggested to add implicit conversion to AD (see https://stackoverflow.com/questions/46720569/implicit-conversion-from-class-to-double)

**Reply by bradbell** [reply]
There is implicit conversion for `double -> AD<double>` because an `AD<double>` can completely represent a `double`. Going from `AD<double>` to `double` looses information that can not be restored; i.e., `AD<double> -> double -> AD<double>` looses information. This is why it is not done implicitly. Perhaps you can accomplishing what you want with template partial specialization for the case 
```
double foo( AD<Scalar> bar )
```

**Reply by bradbell** [reply]
The Var2Par documentation and example now distinguish constant parameter and dynamic parameters; see 06-22 on
https://coin-or.github.io/CppAD/doc/whats_new_21.htm

**Reply by condor-cp** [reply]
Ok I'll do specialization then, thank you
